### ISSUE TITLE
Can't load json-ld document into Arches after exporting from Arches

### ISSUE THREAD

**Issue by apeters** [archesproject/arches#7542]
To reproduce the issue:

1. Create a model with card of cardinality "n" (can be semantic node)
2. Create a sub-card of that card of cardinality "1" (can be semantic node)
3. To that card add a resource-instance node 
4. Create an instance of that model and add two instances of the card
5. Each instance of the resource-instance node should reference the same instance
6. Output the resource instance as json-ld
7. Import that same instance
8. Observe that the import fails with the error `Multiple possible branches and no @type given`

The output of the json-ld might look something like this:
```
{
    "@id": "http://localhost:8000/resources/8758c97c-58e9-480b-b419-cdd41c95432b",
    "@type": "http://www.cidoc-crm.org/cidoc-crm/E29_Design_or_Procedure",
    "http://www.cidoc-crm.org/cidoc-crm/P102_has_title": [
        {
            "@id": "http://localhost:8000/tile/e365f3bd-38fc-4802-b35e-198bde193b25/node/5ce54492-d90c-11eb-97df-acde48001122",
            "@type": "http://www.cidoc-crm.org/cidoc-crm/E35_Title",
            "http://www.cidoc-crm.org/cidoc-crm/P103_was_intended_for": {
                "@id": "http://localhost:8000/tile/e365f3bd-38fc-4802-b35e-198bde193b25/node/788f744c-d90c-11eb-97df-acde48001122",
                "@type": "http://www.cidoc-crm.org/cidoc-crm/E99_Product_Type",
                "http://www.cidoc-crm.org/cidoc-crm/P129i_is_subject_of": {
                    "@id": "http://localhost:8000/resources/5a485b0c-91f8-4ced-8eec-3b137b99bba8"
                }
            }
        },
        {
            "@id": "http://localhost:8000/tile/297625fc-a1ea-47e7-ba44-d1a22ec28553/node/5ce54492-d90c-11eb-97df-acde48001122",
            "@type": "http://www.cidoc-crm.org/cidoc-crm/E35_Title",
            "http://www.cidoc-crm.org/cidoc-crm/P103_was_intended_for": {
                "@id": "http://localhost:8000/tile/297625fc-a1ea-47e7-ba44-d1a22ec28553/node/788f744c-d90c-11eb-97df-acde48001122",
                "@type": "http://www.cidoc-crm.org/cidoc-crm/E99_Product_Type",
                "http://www.cidoc-crm.org/cidoc-crm/P129i_is_subject_of": {
                    "@id": "http://localhost:8000/resources/5a485b0c-91f8-4ced-8eec-3b137b99bba8",
                    "@type": "http://www.cidoc-crm.org/cidoc-crm/E7_Activity"
                }
            }
        }
    ]
}
```

Notice how the `P129i_is_subject_of` value omits the `@type` property on one of the instances because it's already been defined elsewhere in the document.

This is what causes the issue when reloading the document into Arches.

**Comment by azaroth42** [comment]
Agreed that is the output format, as it comes out of the JSON-LD framing algorithm.

This is not limited to the very specific case you've encountered it in.  Another scenario would be two resource-instance references to a Person, one time as the artist, and one time as the thing being depicted in the artwork ... e.g. a self-portrait.  As there are two references to the same entity in the same document, the compaction algorithm would flatten them and drop the `@type` from the second and subsequent -- thus the uri-only form can appear anywhere, and the full reference also anywhere.

Agree with the solution in #7546, although it seems expensive for large documents to always walk the entire tree twice (once to pre-cache the references, once to do the import). Also, that it will store *every* node, even ones that can never be referred to from another part of the graph such as tiles (suggest testing for tiles and not caching them).

Another solution would be to add the `@type` field to the *output* document. This would make the result more consistent, as any client would currently need to do the same thing as in the PR, not just the import code. This would mean walking the tree before serialization before returning it here https://github.com/archesproject/arches/blob/master/arches/app/utils/data_management/resources/formats/rdffile.py#L266 and doing the same sort of test ... cache all dicts that have an `@id` but no `@type`, and when the `@type` is found, update all of the dicts. (Relying on the by-reference-ness of dicts in python, this would update in the main tree).

**Comment by azaroth42** [comment]
BTW, this is also #4563 ;)

**Comment by apeters** [comment]
fixed by #7546
